[HUDI-3921] Fixed schema evolution cannot work with HUDI-3855

xiarixiaoyao opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contribute/how-to-contribute before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds quick-start document.)*

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

xiarixiaoyao commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -431,6 +431,18 @@ public static GenericRecord rewriteRecordWithMetadata(GenericRecord genericRecor
+  // TODO Unify the logical of rewriteRecordWithMetadata and rewriteEvolutionRecordWithMetadata, and delete this function.
+  public static GenericRecord rewriteEvolutionRecordWithMetadata(GenericRecord genericRecord, Schema newSchema, String fileName) {
+    GenericRecord newRecord = HoodieAvroUtils.rewriteRecordWithNewSchema(genericRecord, newSchema, new HashMap<>());
+    // do not preserve FILENAME_METADATA_FIELD
```
just copy from rewriteRecordWithMetadata, 
but i donnot know why we need rewrite genericRecord,  it will cost some time.
can we modfiy genericRecord directly ? just like genericRecord.put(HoodieRecord.FILENAME_METADATA_FIELD_POS, fileName);

xiarixiaoyao replied:
thanks.  After this operation, we will write parquet files directly， The life cycle of genericRecord has come to an end. I think we can try to turn these records into mutable in this place. Of course, let me try this in the next version.

alexeykudinkin replied:
These records might actually be used upstream in some follow-up operations, hence it's preferred to keep them immutable since at this level we don't control their lifecycle

nsivabalan replied:
yes, lets revisit after 0.11 to see if we can avoid full rewrite in some cases. I understand the intent to recreate new record to avoid mutations, but it does incur perf hits. I should have thought about this when we fixed the HoodieMergeHandle for commit time fix in earlier patch. missed to bring it up.

xiarixiaoyao commented:
@alexeykudinkin @bvaradar @xushiyan 
could you pls help me review this pr？  thanks

alexeykudinkin commented:
@xiarixiaoyao left some comments. 

Can you please add a description for this PR? There's very little context in this PR itself, but also not a lot of in the Jira issue, so hard to understand how exactly HUDI-3855 is related to it.

xiarixiaoyao commented:
@alexeykudinkin 
Thank you very much for your review， addressed all comments
add more test for nested rename operation.

by HUDI-3855: we will rewrite old record before write it to parquet file
for schema evolution rename  scene， since old parquet file has old name， when we rewrite the old record with new schema， the value belong to old name will be missed  which lead to a serious problem
for example；
1）now current cow hoodie table have a old parquet file which schema is： a int, b string
2) we rename  a -> aa,  now new schema for hoodie table  will be :  aa int, b string
3) let us insert new data to current hoodie table,  during the insert operation we need to read old record from old parquet file,
**before HUDI-3855**:  we can read old record directly and write it to new parquet directly, rename operation has nothing influence to it
**after HUDI-3855**: before we write old record, we need rewrite it with new schema,  now the schema of old record is: a int, b string but the new schema is: aa int, b string,  if we rewrite the old record forcely we will miss the value of column a since it is not exists in new schema.

xiarixiaoyao commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieWriteHandle.java:
```diff
@@ -98,6 +101,8 @@
   protected final String writeToken;
   protected final TaskContextSupplier taskContextSupplier;
+  // For full schema evolution
+  protected final boolean schemaOnReadEnable;
```
fixed

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -405,6 +407,18 @@ public static GenericRecord rewriteRecordWithMetadata(GenericRecord genericRecor
+    GenericRecord newRecord = HoodieAvroUtils.rewriteRecordWithNewSchema(genericRecord, newSchema, new HashMap<>());
+    // do not preserve FILENAME_METADATA_FIELD
+    newRecord.put(HoodieRecord.FILENAME_METADATA_FIELD_POS, fileName);
+    if (!GenericData.get().validate(newSchema, newRecord)) {
```
won't this take a perf hit if we validate schema compatability for every record? can't we move this outside and do it once for just one of the record?

xiarixiaoyao replied:
this function is copy from rewriteRecordWithMetadata，only used for schema evolution。maybe we can remove this check diectly，we need also do the same things for rewriteRecordWithMetadata

nsivabalan replied:
lets revisit this after 0.11. do not want to drag 0.11 anymore.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/HoodieMergeHelper.java:
```diff
@@ -109,10 +113,14 @@ public void runMerge(HoodieTable<T, HoodieData<HoodieRecord<T>>, HoodieData<Hood
       Schema writeSchemaFromFile = AvroInternalSchemaConverter.convert(writeInternalSchema, readSchema.getName());
       needToReWriteRecord = sameCols.size() != colNamesFromWriteSchema.size()
               || SchemaCompatibility.checkReaderWriterCompatibility(writeSchemaFromFile, readSchema).getType() == org.apache.avro.SchemaCompatibility.SchemaCompatibilityType.COMPATIBLE;
+      if (needToReWriteRecord) {
```
not exactly related to this patch. but in L120, are we passing the arguements in right order? from the docs(SchemaCompatibility.checkReaderWriterCompatibility()), first arg refers to reader schema and 2nd arg refers to writer schema. 
can you check that once please?

xiarixiaoyao replied:
thanks for your remind，let me fixed it

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -741,10 +769,23 @@ private static Object rewriteRecordWithNewSchema(Object oldRecord, Schema oldSch
           if (oldSchema.getField(field.name()) != null) {
             Schema.Field oldField = oldSchema.getField(field.name());
-            helper.put(i, rewriteRecordWithNewSchema(indexedRecord.get(oldField.pos()), oldField.schema(), fields.get(i).schema()));
+            helper.put(i, rewriteRecordWithNewSchema(indexedRecord.get(oldField.pos()), oldField.schema(), fields.get(i).schema(), renameCols, fieldNames));
```
I am bit skeptical on making such changes at this stage of the release. Can we add a new method when renameCols is not empty and not touch existing code (if schema evol is not enabled). Atleast I want to make sure we don't cause any unintentional regression for non schema evol code path.

xiarixiaoyao replied:
donot worry about it。this function is only used by schema evolution，so it will be safe to modify it directly

nsivabalan replied:
thanks!

xiarixiaoyao commented:
@nsivabalan  @alexeykudinkin  could you pls review again

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -741,10 +765,23 @@ private static Object rewriteRecordWithNewSchema(Object oldRecord, Schema oldSch
+          } else {
+            String fieldFullName = createFullName(fieldNames);
+            String[] colNamePartsFromOldSchema = renameCols.getOrDefault(fieldFullName, "").split("\\.");
+            String lastColNameFromOldSchema = colNamePartsFromOldSchema[colNamePartsFromOldSchema.length - 1];
```
nit: `fieldNameFromOldSchema`

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -741,10 +765,23 @@ private static Object rewriteRecordWithNewSchema(Object oldRecord, Schema oldSch
 
         for (int i = 0; i < fields.size(); i++) {
```
Why not just iterating over `fields` themselves?

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -741,10 +765,23 @@ private static Object rewriteRecordWithNewSchema(Object oldRecord, Schema oldSch
+          fieldNames.pop();
         }
         GenericData.Record newRecord = new GenericData.Record(newSchema);
         for (int i = 0; i < fields.size(); i++) {
```
Please check my comment above, we can do this while we iterate over the fields directly to do it in a single loop

xiarixiaoyao replied:
yes , already rework those codes

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -765,27 +802,41 @@ private static Object rewriteRecordWithNewSchema(Object oldRecord, Schema oldSch
         }
         Map<Object, Object> map = (Map<Object, Object>) oldRecord;
         Map<Object, Object> newMap = new HashMap<>();
+        fieldNames.push("value");
```
Please make all these static constants

xiarixiaoyao replied:
fixed

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -765,27 +802,41 @@ private static Object rewriteRecordWithNewSchema(Object oldRecord, Schema oldSch
+    String result = "";
+    if (!fieldNames.isEmpty()) {
+      List<String> parentNames = new ArrayList<>();
+      fieldNames.descendingIterator().forEachRemaining(parentNames::add);
```
You don't need additional list, you can just iterate over deque itself

xiarixiaoyao replied:
fixed

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -379,7 +380,7 @@ private void processDataBlock(HoodieDataBlock dataBlock, Option<KeySpec> keySpec
       while (recordIterator.hasNext()) {
         IndexedRecord currentRecord = recordIterator.next();
-        IndexedRecord record = schemaOption.isPresent() ? HoodieAvroUtils.rewriteRecordWithNewSchema(currentRecord, schemaOption.get()) : currentRecord;
+        IndexedRecord record = schemaOption.isPresent() ? HoodieAvroUtils.rewriteRecordWithNewSchema(currentRecord, schemaOption.get(), new HashMap<>()) : currentRecord;
```
Same comment regarding `HashMap`

alexeykudinkin replied:
Ditto here and everywhere

xiarixiaoyao replied:
yes, fixed

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/internal/schema/action/InternalSchemaMerger.java:
```diff
@@ -48,6 +48,25 @@ public class InternalSchemaMerger {
+  // we can pass colNewName to reWriteRecordWithNewSchema directly, everything is ok.
+  private boolean useColNameFromFileSchema = true;
+
+  public InternalSchemaMerger(InternalSchema fileSchema, InternalSchema querySchema, boolean ignoreRequiredAttribute, boolean useColumnTypeFromFileSchema, boolean useColNameFromFileSchema) {
```
Please chain ctors (ie one invokes another), to avoid duplication

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/internal/schema/action/InternalSchemaMerger.java:
```diff
@@ -131,12 +150,15 @@ private List<Types.Field> buildRecordType(List<Types.Field> oldFields, List<Type
     if (newType.isNestedType()) {
-      return Types.Field.get(oldField.fieldId(), oldField.isOptional(), nameFromFileSchema, newType, oldField.doc());
+      return Types.Field.get(oldField.fieldId(), oldField.isOptional(),
+          useColNameFromFileSchema ? nameFromFileSchema : nameFromQuerySchema, newType, oldField.doc());
```
Please inline as a var and reuse

xiarixiaoyao replied:
fixed

alexeykudinkin commented on hudi-common/src/main/java/org/apache/hudi/internal/schema/utils/InternalSchemaUtils.java:
```diff
@@ -267,4 +267,20 @@ public static String createFullName(String name, Deque<String> fieldNames) {
+    return colNamesFromWriteSchema.stream().filter(f -> {
+      int filedIdFromWriteSchema = oldSchema.findIdByName(f);
+      // try to find the cols which has the same id, but have different colName;
+      return newSchema.getAllIds().contains(filedIdFromWriteSchema) && !newSchema.findfullName(filedIdFromWriteSchema).equalsIgnoreCase(f);
```
Instead of duplicating the code just do a `map` first, where you map the name if it's a rename, otherwise return null, then filter all nulls

alexeykudinkin commented on hudi-common/src/test/java/org/apache/hudi/internal/schema/utils/TestAvroSchemaEvolutionUtils.java:
```diff
@@ -284,7 +284,7 @@ public void testReWriteRecordWithTypeChanged() {
     InternalSchema newSchema = SchemaChangeUtils.applyTableChanges2Schema(internalSchema, updateChange);
     Schema newAvroSchema = AvroInternalSchemaConverter.convert(newSchema, avroSchema.getName());
-    GenericRecord newRecord = HoodieAvroUtils.rewriteRecordWithNewSchema(avroRecord, newAvroSchema);
+    GenericRecord newRecord = HoodieAvroUtils.rewriteRecordWithNewSchema(avroRecord, newAvroSchema, new HashMap<>());
```
Same comment as above

xiarixiaoyao replied:
fixed

alexeykudinkin commented on hudi-common/src/test/java/org/apache/hudi/internal/schema/utils/TestAvroSchemaEvolutionUtils.java:
```diff
@@ -349,7 +349,7 @@ public void testReWriteNestRecord() {
 
     Schema newAvroSchema = AvroInternalSchemaConverter.convert(newRecord, schema.getName());
-    GenericRecord newAvroRecord = HoodieAvroUtils.rewriteRecordWithNewSchema(avroRecord, newAvroSchema);
+    GenericRecord newAvroRecord = HoodieAvroUtils.rewriteRecordWithNewSchema(avroRecord, newAvroSchema, new HashMap<>());
```
Here as well

xiarixiaoyao replied:
fixed

alexeykudinkin commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/TestSpark3DDL.scala:
```diff
@@ -445,28 +445,19 @@ class TestSpark3DDL extends TestHoodieSqlBase {
             Seq(null),
             Seq(Map("t1" -> 10.0d))
           )
+          spark.sql(s"alter table ${tableName} rename column members to mem")
```
Let's in addition to these ones add tests for record rewriting utils in `HoodieAvroUtils`

xiarixiaoyao replied:
thanks, already added

alexeykudinkin reviewed (approved):
LGTM

@xiarixiaoyao i left a bunch of comments, would be great if you could follow up on them in a separate PR

nsivabalan commented:
@xiarixiaoyao : please address the feedback in a follow up PR. I am going ahead and landing this.

xiarixiaoyao commented:
@nsivabalan @alexeykudinkin  thanks for your review, let me put another pr to optimize the code

xiarixiaoyao commented:
@alexeykudinkin @nsivabalan  fixed all comments  on https://github.com/apache/hudi/pull/5393/files

xiarixiaoyao commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -719,14 +729,28 @@ public static Object getRecordColumnValues(HoodieRecord<? extends HoodieRecordPa
-  public static GenericRecord rewriteRecordWithNewSchema(IndexedRecord oldRecord, Schema newSchema) {
-    Object newRecord = rewriteRecordWithNewSchema(oldRecord, oldRecord.getSchema(), newSchema);
+  public static GenericRecord rewriteRecordWithNewSchema(IndexedRecord oldRecord, Schema newSchema, Map<String, String> renameCols) {
+    Object newRecord = rewriteRecordWithNewSchema(oldRecord, oldRecord.getSchema(), newSchema, renameCols, new LinkedList<>());
```
fixed

xiarixiaoyao commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -405,6 +407,14 @@ public static GenericRecord rewriteRecordWithMetadata(GenericRecord genericRecor
 
+  // TODO Unify the logical of rewriteRecordWithMetadata and rewriteEvolutionRecordWithMetadata, and delete this function.
+  public static GenericRecord rewriteEvolutionRecordWithMetadata(GenericRecord genericRecord, Schema newSchema, String fileName) {
+    GenericRecord newRecord = HoodieAvroUtils.rewriteRecordWithNewSchema(genericRecord, newSchema, new HashMap<>());
```
fixed

xiarixiaoyao replied:
fixed
